homeassistant/helpers/entity.py

```diff
@@ -218,8 +219,134 @@ class EntityPlatformState(Enum):
     REMOVED = auto()
 
 
-@dataclass(slots=True)
-class EntityDescription:
+def _class_fields(cls: type, kw_only: bool) -> list[tuple[str, Any, Any]]:
+    """Return a list of dataclass fields.
+
+    Extracted from dataclasses._process_class.
+    """
+    # pylint: disable=protected-access
+    cls_annotations = cls.__dict__.get("__annotations__", {})
+
+    cls_fields: list[dataclasses.Field[Any]] = []
+
+    _dataclasses = sys.modules[dataclasses.__name__]
+    for name, _type in cls_annotations.items():
+        # See if this is a marker to change the value of kw_only.
+        if dataclasses._is_kw_only(type, _dataclasses) or (  # type: ignore[attr-defined]
+            isinstance(_type, str)
+            and dataclasses._is_type(  # type: ignore[attr-defined]
+                _type,
+                cls,
+                _dataclasses,
+                dataclasses.KW_ONLY,
+                dataclasses._is_kw_only,  # type: ignore[attr-defined]
+            )
+        ):
+            kw_only = True
+        else:
+            # Otherwise it's a field of some type.
+            cls_fields.append(dataclasses._get_field(cls, name, _type, kw_only))  # type: ignore[attr-defined]
+
+    return [(field.name, field.type, field) for field in cls_fields]
+
+
+class FrozenOrThawed(type):
```

Comment:
Maybe `FrozenOrThawed` and  `_class_fields` this into a another module and make a note its temporary until we enforce the frozen dataclass requirement

Reply:
Makes sense, move to homeassistant/util/frozen_dataclass_compat.py